batch/batch/driver/instance_collection/base.py

```diff
@@ -87,6 +96,15 @@ def get_instance(self, inst_name) -> Optional[Instance]:
             return inst_coll.name_instance.get(inst_name)
         return None
 
+    async def get_token_from_instance_name(self, name):
+        record: Dict[str, Any] = await self.db.select_and_fetchone(
+            'SELECT token FROM instances WHERE name = %s',
+            (name),
+        )
+        if not record or 'token' not in record:
+            return None
```

Comment:
Hm, there is a type error here that reveals a sticky question. The cache has a value type of `str`, which `None` is not. Should it remain `str` or be `Optional[str]`? Is `None` a value that we *want* in the cache? Would I be able to submit a bunch of garbage instance names and essentially purge the cache of all useful values?

I think we want to guarantee here that the `load` function we give to our cache will produce a valid result. i.e. we should be able to assert that the record we retrieve from the database is not None. Can we do that?

Reply:
I think we would have to change the logic in the TimeLimitedMaxSizeCache.lookup function. Currently, it awaits the result of calling load, puts the result in the cache, and then returns the result. It looks like it doesn't handle the case where a user looks up an item that isn't present in the data source

Reply:
You're right about the behavior of `lookup`, but our *need* for this behavior isn't apparent to me, at least not yet. Based on our usage of this cache, it feels over-protective to think that `record` could be None. Since we keep old workers in the database, `record` should be not None so long as `name` is a real name of an instance that existed. If you look at `active_instances_only`, we throw unauthorized if a request submits an instance name that does not exist. By the time we get to using the cache, we've created an invariant that the input we've received is a good one, and we should take advantage of it! I think we can simply assert here that the record isn't None.